p2p: Refactor sock to add I2P fuzz and unit tests

vasild opened the pull request:
Change the networking code and the I2P code to be fully mockable and use `FuzzedSocket` to fuzz the I2P methods `Listen()`, `Accept()` and `Connect()`.

Add a mocked `Sock` implementation that returns a predefined data on reads and use it for a regression unit test for the bug fixed in https://github.com/bitcoin/bitcoin/pull/21407.

vasild committed:
i2p: use pointers to Sock to accommodate mocking

Change the types of `i2p::Connection::sock` and
`i2p::sam::Session::m_control_sock` from `Sock` to
`std::unique_ptr<Sock>`.

Using pointers would allow us to sneak `FuzzedSock` instead of `Sock`
and have the methods of the former called.

After this change a test only needs to replace `CreateSock()` with
a function that returns `FuzzedSock`.

vasild committed:
fuzz: add tests for the I2P Session public interface

vasild committed:
fuzz: implement unimplemented FuzzedSock methods

We want `Get()` to always return the same value, otherwise it will look
like the `FuzzedSock` implementation itself is broken. So assign
`m_socket` a random number in the `FuzzedSock` constructor.

There is nothing to fuzz about the `Get()` and `Release()` methods, so
use the ones from the base class `Sock`.

`Reset()` is just setting our socket to `INVALID_SOCKET`. We don't want
to use the base `Reset()` because it will close `m_socket` and given
that our `m_socket` is just a random number it may end up closing a real
opened file descriptor if it coincides with our random `m_socket`.

vasild committed:
fuzz: extend FuzzedSock::Recv() to support MSG_PEEK

A conforming `recv(2)` call is supposed to return the same data on a
call following `recv(..., MSG_PEEK)`. Extend `FuzzedSock::Recv()` to do
that.

For simplicity we only return 1 byte when `MSG_PEEK` is used. If we
would return a buffer of N bytes, then we would have to keep track how
many of them were consumed on subsequent non-`MSG_PEEK` calls.

vasild committed:
net: add connect() and getsockopt() wrappers to Sock

Extend the `Sock` class with wrappers to `connect()` and `getsockopt()`.

This will make it possible to mock code which uses those.

vasild committed:
net: change ConnectSocketDirectly() to take a Sock argument

Change `ConnectSocketDirectly()` to take a `Sock` argument instead of a
bare `SOCKET`. With this, use the `Sock`'s (possibly mocked) methods
`Connect()`, `Wait()` and `GetSockOpt()` instead of calling the OS
functions directly.

vasild committed:
fuzz: avoid FuzzedSock::Recv() repeated errors with EAGAIN

If `recv(2)` returns an error (`-1`) and sets `errno` to a temporary
error like `EAGAIN` a proper application code is expected to retry the
operation.

If the fuzz data is exhausted, then `FuzzedSock::Recv()` will keep
returning `-1` and setting `errno` to the first element of
`recv_errnos[]` which happened to be `EAGAIN`. This may continue forever
or cause the fuzz test to run for a long time before some higher level
application "receive timeout" is triggered.

Thus, put `ECONNREFUSED` as first element of `recv_errnos[]`.

jonatack commented:
Concept ACK

vasild commented on src/netbase.cpp:
```diff
@@ -642,46 +642,34 @@ bool ConnectSocketDirectly(const CService &addrConnect, const SOCKET& hSocket, i
             // Connection didn't actually fail, but is being established
             // asynchronously. Thus, use async I/O api (select/poll)
             // synchronously to check for successful connection with a timeout.
-#ifdef USE_POLL
-            struct pollfd pollfd = {};
-            pollfd.fd = hSocket;
-            pollfd.events = POLLIN | POLLOUT;
-            int nRet = poll(&pollfd, 1, nTimeout);
-#else
-            struct timeval timeout = MillisToTimeval(nTimeout);
-            fd_set fdset;
-            FD_ZERO(&fdset);
-            FD_SET(hSocket, &fdset);
-            int nRet = select(hSocket + 1, nullptr, &fdset, nullptr, &timeout);
-#endif
```
Note to reviewers: this old code was inconsistent with the events it requested depending on which function was used - it requested "in" or "out" (`POLLIN | POLLOUT`) when using `poll()` but only "out" when using `select()` (did not pass `fdset` as 2'nd arg to `select()`).

This is now replaced by `Sock::Wait(..., requested Sock::RECV | Sock::SEND)` which, if ends up calling `select()`, will ask it for both "in" and "out".

I think this is ok.

MarcoFalke replied:
it might be good to split the non-fuzz "refactors" out into their own pull

vasild replied:
Lets see which are the non-fuzz refactors:

```
[7] i2p: add fuzz tests on the I2P Session public interface
[6] i2p: use pointers to Sock to accommodate mocking
[5] net: change ConnectSocketDirectly() to take a Sock argument
[4] net: add connect() and getsockopt() wrappers to Sock
[3] fuzz: avoid FuzzedSock::Recv() repeated errors with EAGAIN
[2] fuzz: extend FuzzedSock::Recv() to support MSG_PEEK
[1] fuzz: implement unimplemented FuzzedSock methods
```

`1-3` are obviously fuzz. `4` touches `FuzzedSocket`. `7` adds fuzz tests. So only `5` and `6` remain. `5` depends on the previous commits. `6` is non-fuzz, does not depend on the previous commits and can be extracted, but `7` depends on it.
So a split can be:
PR-A: 6
PR-B: 1-5, 7

I think it is not worth to split, given that 6 is a small mechanical change.

MarcoFalke commented on src/util/sock.h:
```diff
@@ -155,6 +155,8 @@ class Sock
      */
     SOCKET m_socket;
+
+    friend class FuzzedSock;
```
style-nit in the first commit: Wouldn't it be better to change `private` to `protected`? This would allow other tests to mock `m_socket` without having to enumerate all "friend" test classes here

vasild replied:
Done.

vasild commented on src/test/fuzz/util.h:
```diff
@@ -534,36 +534,37 @@ class FuzzedSock : public Sock
+     * If `MSG_PEEK` is used, then our `Recv()` returns some random data as usual, but on the next
+     * `Recv()` call we must return the same data, thus we remember it here.
+     */
+    mutable std::optional<uint8_t> m_peek_data;
```
Either this should be `mutable` or the `const` qualifier should be removed from `Sock::Recv()` (and `Sock::Send()` for consistency). I don't have a strong opinion. I chose `mutable` because it is a smaller change and did not require modifying existing code.

vasild replied:
Removing `const` is not feasible (an avalanche of changes, not worth it).

practicalswift commented:
Concept ACK

Thanks for improving `FuzzedSock` with `connect`, `getsockopt` and `MSG_PEEK` (`recv`)  support! :)

Very happy to see the I2P code being thoroughly fuzzed before landing in a release! "Fuzz before release" is a nice high bar I think we should try to aim for going forward for new features :)

jonatack commented on src/test/fuzz/util.h:
```diff
@@ -642,6 +664,49 @@ class FuzzedSock : public Sock
+            return 0;
+        }
+        const auto rnd = m_fuzzed_data_provider.ConsumeBytes<uint8_t>(*opt_len);
+        std::memcpy(opt_val, rnd.data(), rnd.size());
```
Builds cleanly at a6c20d2b88ea92, but I'm seeing a UBSan warning when starting the fuzzer on each of 3 times:

 ```
$ clang --version
clang version 9.0.1-16 
Target: x86_64-pc-linux-gnu
Thread model: posix

$ FUZZ=i2p src/test/fuzz/fuzz ../qa-assets/fuzz_seed_corpus/
INFO: Seed: 3673367385
INFO: Loaded 1 modules   (642733 inline 8-bit counters): 642733 [0x55abe383e4c8, 0x55abe38db375), 
INFO: Loaded 1 PC tables (642733 PCs): 642733 [0x55abe38db378,0x55abe42a9e48), 
INFO:   132706 files found in ../qa-assets/fuzz_seed_corpus/
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 1048576 bytes
INFO: seed corpus: files: 132706 min: 1b max: 3986616b total: 1882953565b rss: 284Mb
#4096	pulse  cov: 1773 ft: 2176 corp: 11/47b exec/s: 2048 rss: 366Mb

test/fuzz/util.h:703:30: runtime error: null pointer passed as argument 2, which is declared to never be null
/usr/include/string.h:44:28: note: nonnull attribute specified here
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior test/fuzz/util.h:703:30 in 

#8192	pulse  cov: 2600 ft: 3813 corp: 32/219b exec/s: 2730 rss: 441Mb
#16384	pulse  cov: 2655 ft: 4890 corp: 70/750b exec/s: 2730 rss: 586Mb
#32768	pulse  cov: 2773 ft: 6429 corp: 111/2011b exec/s: 2730 rss: 649Mb
#65536	pulse  cov: 2794 ft: 7344 corp: 142/4922b exec/s: 2849 rss: 649Mb
#131072	pulse  cov: 2890 ft: 8189 corp: 162/47Kb exec/s: 2473 rss: 1271Mb
#132707	INITED cov: 2890 ft: 8189 corp: 162/47Kb exec/s: 2249 rss: 1271Mb
#132896	REDUCE cov: 2890 ft: 8189 corp: 162/46Kb lim: 14309 exec/s: 2214 rss: 1271Mb L: 13800/13800 MS: 4 ChangeBinInt-ChangeByte-ChangeByte-EraseBytes-
.../...
#204032	REDUCE cov: 2912 ft: 9195 corp: 194/29Kb lim: 14413 exec/s: 2147 rss: 1271Mb L: 62/2551 MS: 2 ChangeBinInt-EraseBytes-
#206627	REDUCE cov: 2912 ft: 9195 corp: 194/29Kb lim: 14426 exec/s: 2152 rss: 1271Mb L: 66/2551 MS: 5 InsertByte-PersAutoDict-ChangeByte-EraseBytes-PersAutoDict- DE: "\x00\x00\x00\x00\x00\x00\x00\x1a"-"\x1a\x00\x00\x00\x00\x00\x00\x00"-
#207638	REDUCE cov: 2912 ft: 9195 corp: 194/29Kb lim: 14426 exec/s: 2162 rss: 1271Mb L: 102/2551 MS: 1 EraseBytes-
```

vasild replied:
Fixed, thanks! I did not see this because my `memcpy()` does not have the nonnull attribute.

Off-topic: the nonnull attribute has this speciality - the compiler assumes that the parameter will never be null and may eliminate branches in the function that check for that. For example:

```cpp
void func(char* buf) __nonnull(1)
{
    if (buf == nullptr) {
        // take action
        return;
    }
    buf[0] = 1;
}
```

Compiler optimization may reduce this to:

```cpp
void func(char* buf) __nonnull(1)
{
    buf[0] = 1;
}
```

It looks reasonable, given that code like `func(nullptr);` will not compile. However, if the value of the argument is determined at runtime and may still end up being `nullptr`, then the compiler has no way to detect this during compilation. So the line `buf[0] = 1;` may still be executed with `buf` being `nullptr` even though the source code checks for that just above, leading to some nice "wtf" moments.

Ubsan alleviates this.

jonatack replied:
Yes. Runs without the error now.

jonatack replied:
Oops, getting another one.

vasild commented:
`a6c20d2b8...933d181a8`: avoid calling `memcpy()` with `NULL` argument even if `size` is `0`

jonatack commented:
Saw this OOM error twice, but only with  `../qa-assets/fuzz_seed_corpus/`

<details><summary>fuzz output</summary><p>

```
$ FUZZ=i2p src/test/fuzz/fuzz ../qa-assets/fuzz_seed_corpus/
INFO: Seed: 2309095707
INFO: Loaded 1 modules   (642728 inline 8-bit counters): 642728 [0x55bbf8e17668, 0x55bbf8eb4510), 
INFO: Loaded 1 PC tables (642728 PCs): 642728 [0x55bbf8eb4510,0x55bbf9882f90), 
INFO:   237105 files found in ../qa-assets/fuzz_seed_corpus/
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 1048576 bytes
INFO: seed corpus: files: 237105 min: 1b max: 3986616b total: 4147898382b rss: 359Mb
#4096	pulse  cov: 1461 ft: 1820 corp: 8/29b exec/s: 2048 rss: 445Mb
#8192	pulse  cov: 1862 ft: 2571 corp: 23/136b exec/s: 2048 rss: 520Mb
#16384	pulse  cov: 2655 ft: 4120 corp: 38/278b exec/s: 2340 rss: 659Mb
#32768	pulse  cov: 2749 ft: 5425 corp: 81/926b exec/s: 2730 rss: 722Mb
#65536	pulse  cov: 2813 ft: 6711 corp: 132/2841b exec/s: 2849 rss: 722Mb
#131072	pulse  cov: 2911 ft: 8428 corp: 169/7173b exec/s: 2730 rss: 722Mb
==16390== ERROR: libFuzzer: out-of-memory (used: 2191Mb; limit: 2048Mb)
   To change the out-of-memory limit use -rss_limit_mb=<N>

Live Heap Allocations: 89301481 bytes in 242997 chunks; quarantined: 245829231 bytes in 30252 chunks; 1909049 other chunks; total chunks: 2182298; showing top 95% (at most 8 unique contexts)
22746576 byte(s) (25%) in 237105 allocation(s)
    #0 0x55bbf4cf94fd in malloc (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2d444fd)
    #1 0x55bbf4c0bde7 in operator new(unsigned long) (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c56de7)
    #2 0x55bbf4c23aac in fuzzer::ReadCorpora(std::Fuzzer::vector<std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> >, fuzzer::fuzzer_allocator<std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > > > const&, std::Fuzzer::vector<std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> >, fuzzer::fuzzer_allocator<std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > > > const&) (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c6eaac)
    #3 0x55bbf4c23672 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c6e672)
    #4 0x55bbf4c4cb12 in main (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c97b12)
    #5 0x7f929ca83d09 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26d09)

21499328 byte(s) (24%) in 12 allocation(s)
    #0 0x55bbf4cf94fd in malloc (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2d444fd)
    #1 0x55bbf4c0bde7 in operator new(unsigned long) (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c56de7)
    #2 0x55bbf4c4cb12 in main (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c97b12)
    #3 0x7f929ca83d09 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26d09)

16777216 byte(s) (18%) in 1 allocation(s)
    #0 0x55bbf4d28c3d in operator new(unsigned long) (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2d73c3d)
    #1 0x55bbf4e228c5 in __gnu_cxx::new_allocator<uint256>::allocate(unsigned long, void const*) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/ext/new_allocator.h:115:27
    #2 0x55bbf4e228c5 in std::allocator_traits<std::allocator<uint256> >::allocate(std::allocator<uint256>&, unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/alloc_traits.h:460:20
    #3 0x55bbf4e228c5 in std::_Vector_base<uint256, std::allocator<uint256> >::_M_allocate(unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/stl_vector.h:346:20
    #4 0x55bbf575d27a in std::vector<uint256, std::allocator<uint256> >::_M_default_append(unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/vector.tcc:635:34
    #5 0x55bbf575cdd2 in std::vector<uint256, std::allocator<uint256> >::resize(unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/stl_vector.h:940:4
    #6 0x55bbf575c623 in CuckooCache::cache<uint256, SignatureCacheHasher>::setup(unsigned int) /home/jon/projects/bitcoin/bitcoin/src/./cuckoocache.h:344:15
    #7 0x55bbf59809e6 in CuckooCache::cache<uint256, SignatureCacheHasher>::setup_bytes(unsigned long) /home/jon/projects/bitcoin/bitcoin/src/./cuckoocache.h:368:16
    #8 0x55bbf59809e6 in InitScriptExecutionCache() /home/jon/projects/bitcoin/bitcoin/src/validation.cpp:1468:44
    #9 0x55bbf607ba7f in BasicTestingSetup::BasicTestingSetup(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<char const*, std::allocator<char const*> > const&) /home/jon/projects/bitcoin/bitcoin/src/test/util/setup_common.cpp:110:5
    #10 0x55bbf4dc2187 in std::_MakeUniq<BasicTestingSetup const>::__single_object std::make_unique<BasicTestingSetup const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<char const*, std::allocator<char const*> > const&>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<char const*, std::allocator<char const*> > const&) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/unique_ptr.h:962:34
    #11 0x55bbf4dbe382 in std::unique_ptr<BasicTestingSetup const, std::default_delete<BasicTestingSetup const> > MakeNoLogFileContext<BasicTestingSetup const>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<char const*, std::allocator<char const*> > const&) /home/jon/projects/bitcoin/bitcoin/src/./test/util/setup_common.h:170:12
    #12 0x55bbf4fe4723 in initialize_i2p() /home/jon/projects/bitcoin/bitcoin/src/test/fuzz/i2p.cpp:17:39
    #13 0x55bbf4d2f80c in void std::__invoke_impl<void, void (*&)()>(std::__invoke_other, void (*&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:60:14
    #14 0x55bbf4d2f80c in std::enable_if<is_invocable_r_v<void, void (*&)()>, void>::type std::__invoke_r<void, void (*&)()>(void (*&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:110:2
    #15 0x55bbf4d2f80c in std::_Function_handler<void (), void (*)()>::_M_invoke(std::_Any_data const&) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/std_function.h:291:9
    #16 0x55bbf531c54c in std::function<void ()>::operator()() const /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/std_function.h:622:14
    #17 0x55bbf66a3c92 in initialize() /home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz.cpp:44:5
    #18 0x55bbf66a51cd in LLVMFuzzerInitialize /home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz.cpp:70:5
    #19 0x55bbf4c218ac in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c6c8ac)
    #20 0x55bbf4c4cb12 in main (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c97b12)
    #21 0x7f929ca83d09 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26d09)

16777216 byte(s) (18%) in 1 allocation(s)
    #0 0x55bbf4d28c3d in operator new(unsigned long) (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2d73c3d)
    #1 0x55bbf4e228c5 in __gnu_cxx::new_allocator<uint256>::allocate(unsigned long, void const*) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/ext/new_allocator.h:115:27
    #2 0x55bbf4e228c5 in std::allocator_traits<std::allocator<uint256> >::allocate(std::allocator<uint256>&, unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/alloc_traits.h:460:20
    #3 0x55bbf4e228c5 in std::_Vector_base<uint256, std::allocator<uint256> >::_M_allocate(unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/stl_vector.h:346:20
    #4 0x55bbf575d27a in std::vector<uint256, std::allocator<uint256> >::_M_default_append(unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/vector.tcc:635:34
    #5 0x55bbf575cdd2 in std::vector<uint256, std::allocator<uint256> >::resize(unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/stl_vector.h:940:4
    #6 0x55bbf575c623 in CuckooCache::cache<uint256, SignatureCacheHasher>::setup(unsigned int) /home/jon/projects/bitcoin/bitcoin/src/./cuckoocache.h:344:15
    #7 0x55bbf5759dc2 in CuckooCache::cache<uint256, SignatureCacheHasher>::setup_bytes(unsigned long) /home/jon/projects/bitcoin/bitcoin/src/./cuckoocache.h:368:16
    #8 0x55bbf5759dc2 in (anonymous namespace)::CSignatureCache::setup_bytes(unsigned long) /home/jon/projects/bitcoin/bitcoin/src/script/sigcache.cpp:80:25
    #9 0x55bbf5759dc2 in InitSignatureCache() /home/jon/projects/bitcoin/bitcoin/src/script/sigcache.cpp:100:36
    #10 0x55bbf607ba6f in BasicTestingSetup::BasicTestingSetup(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<char const*, std::allocator<char const*> > const&) /home/jon/projects/bitcoin/bitcoin/src/test/util/setup_common.cpp:109:5
    #11 0x55bbf4dc2187 in std::_MakeUniq<BasicTestingSetup const>::__single_object std::make_unique<BasicTestingSetup const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<char const*, std::allocator<char const*> > const&>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<char const*, std::allocator<char const*> > const&) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/unique_ptr.h:962:34
    #12 0x55bbf4dbe382 in std::unique_ptr<BasicTestingSetup const, std::default_delete<BasicTestingSetup const> > MakeNoLogFileContext<BasicTestingSetup const>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<char const*, std::allocator<char const*> > const&) /home/jon/projects/bitcoin/bitcoin/src/./test/util/setup_common.h:170:12
    #13 0x55bbf4fe4723 in initialize_i2p() /home/jon/projects/bitcoin/bitcoin/src/test/fuzz/i2p.cpp:17:39
    #14 0x55bbf4d2f80c in void std::__invoke_impl<void, void (*&)()>(std::__invoke_other, void (*&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:60:14
    #15 0x55bbf4d2f80c in std::enable_if<is_invocable_r_v<void, void (*&)()>, void>::type std::__invoke_r<void, void (*&)()>(void (*&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:110:2
    #16 0x55bbf4d2f80c in std::_Function_handler<void (), void (*)()>::_M_invoke(std::_Any_data const&) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/std_function.h:291:9
    #17 0x55bbf531c54c in std::function<void ()>::operator()() const /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/std_function.h:622:14
    #18 0x55bbf66a3c92 in initialize() /home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz.cpp:44:5
    #19 0x55bbf66a51cd in LLVMFuzzerInitialize /home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz.cpp:70:5
    #20 0x55bbf4c218ac in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c6c8ac)
    #21 0x55bbf4c4cb12 in main (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c97b12)
    #22 0x7f929ca83d09 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26d09)

8388608 byte(s) (9%) in 1 allocation(s)
    #0 0x55bbf4cf94fd in malloc (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2d444fd)
    #1 0x55bbf4c0bde7 in operator new(unsigned long) (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c56de7)
    #2 0x55bbf4c2f807 in fuzzer::GetSizedFilesFromDir(std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > const&, std::Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >*) (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c7a807)
    #3 0x55bbf4c23aac in fuzzer::ReadCorpora(std::Fuzzer::vector<std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> >, fuzzer::fuzzer_allocator<std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > > > const&, std::Fuzzer::vector<std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> >, fuzzer::fuzzer_allocator<std::Fuzzer::basic_string<char, std::Fuzzer::char_traits<char>, std::Fuzzer::allocator<char> > > > const&) (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c6eaac)
    #4 0x55bbf4c23672 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c6e672)
    #5 0x55bbf4c4cb12 in main (/home/jon/projects/bitcoin/bitcoin/src/test/fuzz/fuzz+0x2c97b12)
    #6 0x7f929ca83d09 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26d09)

MS: 0 ; base unit: 0000000000000000000000000000000000000000


artifact_prefix='./'; Test unit written to ./oom-da39a3ee5e6b4b0d3255bfef95601890afd80709
Base64: 
SUMMARY: libFuzzer: out-of-memory
```
</p></details>

vasild committed:
test: add I2P test for a runaway SAM proxy

Add a regression test for https://github.com/bitcoin/bitcoin/pull/21407.

The test creates a socket that, upon read, returns some data, but never
the expected terminator `\n`, injects that socket into the I2P code and
expects `i2p::sam::Session::Connect()` to fail, printing a specific
error message to the log.

jonatack commented:
Had the same OOM issue today with another fuzz PR, so I'm proceeding on the assumption that it's a regression that is not related to this PR.

jonatack reviewed (commented):
Built and reviewed each commit up to 34c199db11b36eee91e56cdd046c090f62da18d6. 

In the 3rd commit, "avoid repeated errors," it looks like the commit message should be "to retry the operation." e.g. s/to/the/

vasild commented:
> Had the same OOM issue today with another fuzz PR

Can you reproduce it? E.g. `FUZZ=i2p src/test/fuzz/fuzz ./oom-da39a3ee5e6b4b0d3255bfef95601890afd80709`

jonatack commented:
```
INFO: Seed: 3949145381
INFO: Loaded 1 modules   (643885 inline 8-bit counters): 643885 [0x56188117c288, 0x5618812195b5), 
INFO: Loaded 1 PC tables (643885 PCs): 643885 [0x5618812195b8,0x561881bec888), 
src/test/fuzz/fuzz: Running 1 inputs 1 time(s) each.
Running: oom-da39a3ee5e6b4b0d3255bfef95601890afd80709
Executed oom-da39a3ee5e6b4b0d3255bfef95601890afd80709 in 0 ms
***
*** NOTE: fuzzing was not performed, you have only
***       executed the target code on a fixed set of inputs.
***
```

jonatack commented on src/i2p.cpp:
```diff
@@ -370,12 +372,12 @@ void Session::CreateIfNotCreatedAlready()
+std::unique_ptr<Sock> Session::StreamAccept()
 {
-    Sock sock = Hello();
+    auto sock = Hello();
```
5ac2bc1 Verified this line and line 352 compile when replaced with explicit typing and braced initialization `std::unique_ptr<Sock> sock{Hello()};` (which would save people time in looking up the type)

jonatack commented on src/test/fuzz/i2p.cpp:
```diff
@@ -0,0 +1,51 @@
+        return std::make_unique<FuzzedSock>(fuzzed_data_provider);
+    };
+
+    CService sam_proxy;
```
933d181a nit, `sam_proxy` here and `to` on line 43 below can both be `const` (to show intention)

vasild replied:
Done.

jonatack reviewed (commented):
ACK 933d181a8c0c41318204b1657765d582418a80a1 modulo a few thoughts/questions above

vasild commented on src/util/sock.h:
```diff
@@ -90,6 +90,19 @@ class Sock
+     * `getsockopt(this->Get(), level, opt_name, opt_val, opt_len)`. Code that uses this
+     * wrapper can be unit-tested if this method is overridden by a mock Sock implementation.
+     */
+    virtual int Getsockopt(int level, int opt_name, void* opt_val, socklen_t* opt_len) const;
```
Done.

vasild commented on src/test/fuzz/util.h:
```diff
@@ -642,6 +664,47 @@ class FuzzedSock : public Sock
         return random_bytes.size();
     }
 
+    int Connect(const sockaddr*, socklen_t) const override
```
Why? It is a bit shorter without the parameter names. Omitting parameter names if they are not used I think is a good practice because it makes it obvious that they are not used.

Also, specifying parameter names and not using those parameters in the function will produce a compiler warning with `-Wunused-parameter` (which is explicitly disabled in Bitcoin Core, I guess because of so many functions that violate it that it is hard to fix).

jonatack replied:
Thanks for the explanation.

vasild commented on src/netbase.cpp:
```diff
@@ -633,7 +633,7 @@ bool ConnectSocketDirectly(const CService &addrConnect, const SOCKET& hSocket, i
 
     // Connect to the addrConnect service on the hSocket socket.
-    if (connect(hSocket, (struct sockaddr*)&sockaddr, len) == SOCKET_ERROR)
+    if (sock.Connect((struct sockaddr*)&sockaddr, len) == SOCKET_ERROR)
```
Done.

vasild commented on src/netbase.cpp:
```diff
@@ -642,46 +642,34 @@ bool ConnectSocketDirectly(const CService &addrConnect, const SOCKET& hSocket, i
+            // sockerr here.
+            int sockerr;
+            socklen_t sockerr_len = sizeof(sockerr);
+            if (sock.Getsockopt(SOL_SOCKET, SO_ERROR, (sockopt_arg_type)&sockerr, &sockerr_len) ==
```
Done.

vasild replied:
Broke the windows build. Reverted.
```
netbase.cpp:663:81: error: invalid static_cast from type ‘int*’ to type ‘sockopt_arg_type’ {aka ‘char*’}
  663 |                     SOL_SOCKET, SO_ERROR, static_cast<sockopt_arg_type>(&sockerr), &sockerr_len) ==
      |                                                                                 ^
```

jonatack replied:
We should probably use `reinterpret_cast` to signal that this is dangerous and non-portable.

vasild commented on src/netbase.cpp:
```diff
@@ -642,46 +642,34 @@ bool ConnectSocketDirectly(const CService &addrConnect, const SOCKET& hSocket, i
-            {
+            // sockerr here.
+            int sockerr;
+            socklen_t sockerr_len = sizeof(sockerr);
```
I don't think so - `getsockopt()` will set it (if it returns `0`). The previous code abused `nRet` which contained a leftover value from the call to `select()`.

vasild replied:
`sizeof()` does not read the value, can be used safely without triggering any ubsan or valgrind warnings on uninitialized variables.

vasild commented:
`933d181a8...7c6fc2de5`: address suggestions and also fuzz the `IsConnected()` method of `Sock` / `FuzzedSock`.

vasild commented:
`7c6fc2de5...23c861d6f`: fix Windows CI

jonatack commented:
re-ACK 23c861d6ff995b7e6034d4bec6af2f6a3d595dca

vasild commented:
`23c861d6f...40316a37c`:

* rebase due to conflicts
* expect that `Sock` methods `RecvUntilTerminator()` and `SendComplete()` could throw an exception when using fuzzed socket. After all a fuzzed socket may return an error at any time.
* add a regression test for the bug fixed in https://github.com/bitcoin/bitcoin/pull/21407.

jonatack reviewed (approved):
Thanks for adding the regression test. 

re-ACK 40316a37cb02cf8a9a8b2cbd4d7153ffa57e7ec5 reviewed `git range-diff 01bb3afb 23c861d 40316a3` and the new unit test commit, debug built, ran unit tests, ran bitcoind with an I2P service and network operation with seven I2P peers (2 in, 5 out) is looking nominal

practicalswift commented:
Tested ACK 40316a37cb02cf8a9a8b2cbd4d7153ffa57e7ec5

Great fuzzing work @vasild!

MarcoFalke commented:
(Changed title, because this is not test-only)

MarcoFalke commented:
Concept ACK 40316a37cb

laanwj reviewed (approved):
Code review ACK 40316a37cb02cf8a9a8b2cbd4d7153ffa57e7ec5

I like how this gets rid of a `#ifdef USE_POLL` in the connect logic.

MarcoFalke commented on src/test/fuzz/i2p.cpp:
```diff
@@ -0,0 +1,57 @@
+
+    if (sess.Connect(to, conn, proxy_error)) {
+        try {
+            conn.sock->SendComplete("verack\n", 10ms, interrupt);
```
Looks like this is never reached?

https://marcofalke.github.io/btc_cov/fuzz.coverage/src/test/fuzz/i2p.cpp.gcov.html

vasild replied:
Well, it would be executed, but for `Connect()` to return `true` the fuzzed sock must return some meaningful data, which I guess has a very very low chance of happening.

MarcoFalke replied:
There'd also be a middle way to teach the fuzz engine a meaningful format or snippets. This can be achieved with a dict, a custom mutator, or with some code in the fuzz target (CallOneOf(...))
